feature(Cashbox for User)

lomamech opened the pull request:
- Implement component bh-multiple-cashbox-select
- Update User registry add in template Action Cell Management Cashbox
- Implement modal to Set Management Cashboxes to User
- Implement API for Set cashbox to User
- Implement Integration test for Cashbox permissions
- Update and fix integration test for Depot Permission
- Implement test E2E and Integration

closes #2242

lomamech committed:
feature(Cashbox for User)
- Implement component bh-multiple-cashbox-select
- Update User registry add in template Action Cell Management Cashbox
- Implement modal to Set Management Cashboxes to User
- Implement API for Set cashbox to User

lomamech committed:
feature(Cash User)
- Implement Integration test for Cashbox permissions
- Update and fix integration test for Depot Permission

lomamech committed:
Fix comment

lomamech committed:
Rebase with Branch master

jniles commented on client/src/i18n/en/cash.json:
```diff
@@ -22,7 +22,9 @@
-      "NO_CURRENT_PROJECT_CASHBOXES" : "No cashboxes found registered to the current project."
+	  "NO_OTHER_PROJECT_CASHBOXES_ACCESS" : "No cashboxes found registered to other projects that you have access right to.",
+      "NO_CURRENT_PROJECT_CASHBOXES" : "No cashboxes found registered to the current project.",
+	  "NO_CURRENT_PROJECT_CASHBOXES_ACCESS" : "There are no cashbox in your project that you have access right to"
```
Nice! :+1:

jniles commented on client/src/js/components/bhCashboxSelect.js:
```diff
@@ -36,7 +36,7 @@ function CashboxSelectController(Cashbox, Notify) {
 
     // load all Cashbox
-    Cashbox.read(null, { detailed: 1 })
+    Cashbox.read(null, { detailed: 1, user_id : Session.user.id })
```
Hmm... the user could in theory alter this, since it's on the client.  I wonder what we should do about that.  We could use `req.session.user.id` on the server, but that would limit ability to download all cashboxes.

We need a better security model in general.  I do not have specific proposals for this route, so this is a non-blocking comment.

jniles commented on client/src/js/components/bhMultipleCashboxSelect.js:
```diff
@@ -0,0 +1,47 @@
+    $ctrl.label = $ctrl.label || 'FORM.LABELS.CASHBOX';
+
+    // fired when a Cashbox has been selected or removed from the list
+    $ctrl.onChange = $ctrl.onChange || angular.noop;
```
Since this function is required in your bindings (`&`), there isn't any need to alias it with a `noop` function.

jniles commented on client/src/modules/cash/modals/select-cashbox-modal.ctrl.js:
```diff
@@ -43,6 +43,27 @@ function SelectCashboxModalController(Session, Instance, Cashboxes, $stateParams
+    /**
+    * This section makes it possible to check if the user does not have permissions to a cash register or that it does not exist
+    */ 
+    Cashboxes.read(undefined, { is_auxiliary : 1 })
```
It would be best not to download the cashboxes twice.  Is there any way to send something like an HTTP parameter `includeUserPrivileges=1`  that would do a `cash_box LEFT JOIN cashbox_permission`?  That way you could loop through the list twice and group by project and user?

jniles commented on client/src/modules/reports/generate/cash_report/cash_report.html:
```diff
@@ -65,24 +65,11 @@ <h3 class="text-capitalize" translate>REPORT.CASH_REPORT.TITLE</h3>
-                </div>
-            </div>
+            <bh-cashbox-select
+              cashboxId = "ReportConfigCtrl.reportDetails.account_id"
```
I think this should be `cashbox-id`.  Also, why is this account id?

lomamech replied:
@jniles, the value to be sent is indeed the account Id, but after analysis this report does not work and there is no End to End test to verify the proper functioning of the report / reports / cash_report

jniles commented on client/src/modules/users/UserCashBoxManagementModal.html:
```diff
@@ -0,0 +1,39 @@
+
+
+    <div class="form-group" ng-class="{ 'has-error' : UserForm.$submitted && UserForm.display_name.$invalid }">
+      <label class="control-label col-md-12" translate>FORM.LABELS.USERNAME</label>
```
Why do you need the `col-md-12` on these lines?

jniles commented on client/src/modules/users/UserCashBoxManagementModal.html:
```diff
@@ -0,0 +1,39 @@
+      </div>
+    </div>
+
+    <span class="help-block" translate> FORM.INFO.USER_CASHBOX </span>
```
I think `help-block`s often go within the `form-group`.  Could you move this up into the form group?

jniles commented on client/src/js/components/bhMultipleCashboxSelect.js:
```diff
@@ -0,0 +1,47 @@
+    controller  : MultipleCashboxSelectController,
+    bindings    : {
+      onChange : '&',
+      cashboxesIds : '<?',
```
This should probably be `cashbox-ids` to make it plural.

jniles commented on client/src/modules/users/users.routes.js:
```diff
@@ -47,7 +47,15 @@ angular.module('bhima.routes')
-      })      
+      })
+      .state('users.cashBoxManagement', {
+        url : '/:id/cashBoxManagement',
```
Since your server URL is ` $http.get('/users/' + id + '/cashboxes')`, it would be nice to make this `/:id/cashboxes`.  That way, it is easier for developers to find it.

jniles commented on server/controllers/admin/users/cashboxes.js:
```diff
@@ -0,0 +1,67 @@
+/**
+ * POST /users/:id/cashboxes
+ *
+ * Creates and updates a user's cashboxes.  This works by completely deleting
```
👍

jniles commented on server/controllers/admin/users/depots.js:
```diff
@@ -49,7 +49,7 @@ function create(req, res, next) {
 
   // if an array of permission has been sent, add them to an INSERT query
-  if (req.body.depots) {
+  if (req.body.depots.length) {
```
💃

jniles commented on server/controllers/finance/cashboxes/index.js:
```diff
@@ -54,6 +54,18 @@ function list(req, res, next) {
   filters.equals('project_id');
   filters.equals('is_auxiliary');
 
+  if (req.query.detailed === '1') {
```
Does just `filters.equals('user_id')` not work for both cases?

lomamech replied:
I used an alternative structure just because in the first case we use the table cash_box_account_currency and in the second it's cash_box

jniles commented on test/data.sql:
```diff
@@ -242,6 +242,13 @@ INSERT INTO `cash_box_account_currency` VALUES
   (4,2,2,3626,3627);
 
+-- Set Cashbox Management By User
+INSERT INTO cashbox_permission (id, user_id, cashbox_id) VALUES
```
Instead of inserting NULLs, you can just do this:
```sql
INSERT INTO cashbox_permission (user_id, cashbox_id) VALUES (1,1), (1,2);
```

jniles commented on test/end-to-end/cash/cash.spec.js:
```diff
@@ -31,6 +35,18 @@ describe('Cash Payments', () => {
   }
 
   describe('Cashbox Select Interface', () => {
+    it(`Set Cashbox ${cashboxC.text} Manage Right to RegularUser `, () => {
```
This belongs in the `users` tests.  It isn't a test tied to cash payments.

jniles commented on test/end-to-end/depots/depots.spec.js:
```diff
@@ -16,12 +16,11 @@ describe('Depots Management', () => {
-  /* user test because the E2E test on the users runs after the repository test. 
+  /* The implementation of the E2E test of the assignment of a Depot
+  /* to a user is added in the E2E test of the Depot and not in the 
+  /* user test because the E2E test on the users runs after the Depot test. 
```
💃

jniles commented on test/integration/users.js:
```diff
@@ -243,4 +244,37 @@ describe('(/users) Users and Permissions', function () {
+  });
+
+  // Reset cashbox permissions for user
+  it('POST /users/:id/cashboxes will reset user cashboxes', function () {
```
👍

jniles commented on test/integration/users.js:
```diff
@@ -243,4 +244,37 @@ describe('(/users) Users and Permissions', function () {
 
+
+  // Add cashbox permissions for user
+  it('POST /users/:id/cashboxes will create user cashboxes', function () {
```
These are really nice tests!

jniles reviewed (commented):
@lomamech, this represents a lot of work.  I'm a bit concerned about the security model employed here ... if a user changes their user_id locally, it will affect the cashboxes they have access to.   I don't have a better proposal right now, but we'll need to come up with one long-term.

You tests are excellent!  They should all be `users` since they are tied to the `users` module.

I've made some code clean up comments.  Let me know what you think.

lomamech committed:
feature(Cashbox for User)
- Remove SessionService user_id in client side by the using of
  req.session.user_id in Server Side
- Sanitaze Multiple Cash box Select Controller remove property
  angular.noop
- Add service Read Privileges for get the privileges for each user for
  each cashboxes
- Update users.routes use /:id/cashboxes instead /:id/cashBoxManagement
- Implemente API for get /cashboxes/privileges
- Set the test for Manage right to Regular user in User E2E test
-

lomamech committed:
test(Cashboxes Privileges)
- Implement integration test for GET /cashboxes/privileges

jniles commented on server/controllers/finance/cashboxes/index.js:
```diff
@@ -36,14 +37,18 @@ exports.currencies = currencies;
  * optional parameters:
  */
 function list(req, res, next) {
+  if (req.query.user_id) {
```
This is pretty confusing.  Why don't you use a variable that isn't a filter property -  Like `showOnlyUserCashboxes`?  That way, developers don't get confused thinking that `user_id = true` is a real user ID somewhere.

jniles reviewed (commented):
@lomamech, one final piece to update, and then this can be merged.

lomamech committed:
refactor(Cash box for User)
- Update cashbox.read use param only_user instead user_id
- Update Depots.read use param only_user instead user_id
- Remove SessionService for the getting of user_id

jniles reviewed (approved):
LGTM!

bors r+
